File: benchmarks/executable/restaurant_uint_saturating_subtract.c
// C23 reference for the Restaurant UInt saturating-subtract workload.

#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

static volatile uint64_t runtime_zero = UINT64_C(0);
static volatile uint64_t runtime_ordinary = UINT64_C(11);
static volatile uint64_t runtime_decrement = UINT64_C(1);

static uint64_t saturating_subtract_u64(uint64_t value, uint64_t amount) {
  if (value < amount) return 0;
  return value - amount;
}

int main(void) {
#ifdef _WIN32
  if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif
  const uint64_t zero =
      saturating_subtract_u64(runtime_zero, runtime_decrement);
  const uint64_t ordinary =
      saturating_subtract_u64(runtime_ordinary, runtime_decrement);
  return printf("Saturated subtract %" PRIu64 "/%" PRIu64 "\n", zero, ordinary) < 0
      ? 1
      : 0;
}

File: benchmarks/executable/restaurant_uint_saturating_subtract.rs
// Rust 2024 reference for the Restaurant UInt saturating-subtract workload.

use std::hint::black_box;
use std::io::{self, Write};

fn saturating_subtract_u64(value: u64, amount: u64) -> u64 {
    value.saturating_sub(amount)
}

fn main() {
    let zero = saturating_subtract_u64(black_box(0_u64), black_box(1_u64));
    let ordinary = saturating_subtract_u64(black_box(11_u64), black_box(1_u64));
    let mut stdout = io::stdout().lock();
    if writeln!(stdout, "Saturated subtract {zero}/{ordinary}").is_err() {
        std::process::exit(1);
    }
}
